src/sinks/prometheus/exporter.rs

```diff
@@ -496,7 +497,7 @@ impl PrometheusExporter {
                 .layer(build_http_trace_layer(span.clone()))
                 .service(inner);
 
-            async move { Ok::<_, Infallible>(service) }
+            async move { Ok::<_, Infallible>(Compression::new(service)) }
```

Comment:
Was getting some errors with the `CompressionLayer` about trait bounds not being satisfied - using `Compression` to wrap the service seemed to make it happier but still need to run remaining tests locally to verify

Reply:
So that we can use `CompressionLayer` here we need to make a number of functions more generic.

`build_http_trace_layer` doesn't need `Body` so we can make those parameters generic. We need separate params for Request and Response. This works:

```rust
pub fn build_http_trace_layer<T, U>(
    span: Span,
) -> TraceLayer<
    SharedClassifier<ServerErrorsAsFailures>,
    impl Fn(&Request<T>) -> Span + Clone,
    impl Fn(&Request<T>, &Span) + Clone,
    impl Fn(&Response<U>, Duration, &Span) + Clone,
    (),
    (),
    (),
>
```

Then `Handler::handle` needs to be generic over `hyper::body::HttpBody` rather than just take `Body`:

```rust
    fn handle<T: HttpBody>(
        &self,
        req: Request<T>,
        metrics: &RwLock<IndexMap<MetricRef, (Metric, MetricMetadata)>>,
    )
```

And similarly `authorized`:

```rust
fn authorized<T: HttpBody>(req: &Request<T>, auth: &Option<Auth>) -> bool
```

Can you make these changes?

Reply:
Thanks for the detail here